### PR TITLE
Fix GBindingGroup bindings

### PR DESCRIPTION
**Pull request by A6GibKm**
Fixes https://github.com/gtk-rs/gtk-rs-core/issues/664.

### CODE REVIEW

**Comment by sdroege** [comment]
CC @jf2048

File: glib/src/gobject/binding_group.rs [hunk 1]
```diff
@@ -140,21 +140,23 @@ impl<'a> BindingGroupBuilder<'a> {
             user_data: ffi::gpointer,
         ) -> ffi::gboolean {
             let transform_data =
-                &*(user_data as *const (TransformFn, TransformFn, ParamSpec, ParamSpec));
+                &*(user_data as *const (TransformFn, TransformFn, Option<ParamSpec>, ParamSpec));
```
Comment (sdroege) [inline glib/src/gobject/binding_group.rs:143]: I don't think this will work like this. You need the source pspec here for checking that the value types are matching, but if there is none duration construction then you can't really get it in here easily.

**Comment by jf2048** [comment]
This should probably get a unit test

File: glib/src/gobject/binding_group.rs (hunk 1 above)
Comment (jf2048) [reply glib/src/gobject/binding_group.rs:143]: This should probably just get rid of the `ParamSpec`s and use strings and `find_property`, that is what the `BindingGroup` does internally

Comment (sdroege) [reply glib/src/gobject/binding_group.rs:143]: Sounds like a plan

**Comment by sdroege** [comment]
What's the status here?

**Comment by A6GibKm** [comment]
> What's the status here?

I been busy this week, I hope to return to it today or tomorrow.

File: glib/src/gobject/binding_group.rs [hunk 2]
```diff
@@ -233,8 +249,157 @@ impl<'a> BindingGroupBuilder<'a> {
...
+        pub struct TestObject {
+            pub name: RefCell<String>,
+            pub enabled: RefCell<bool>,
+        }
+
+        #[crate::object_subclass]
+        impl ObjectSubclass for TestObject {
+            const NAME: &'static str = "Test";
+            type Type = super::TestObject;
+            type ParentType = crate::Object;
```
Comment (bilelmoussaoui) [inline glib/src/gobject/binding_group.rs:319]:
```suggestion
```

**Comment by sdroege** [comment]
@A6GibKm The problem was in your code: you were not NUL-terminating the string if there was no source object.

Now the test is still failing but that's something plainly on the Rust side of things :)

File: glib/src/gobject/binding_group.rs [hunk 3]
```diff
@@ -233,8 +252,159 @@ impl<'a> BindingGroupBuilder<'a> {
...
+
+        source.set_name("test_source_name");
+        assert_eq!(source.name(), target.name());
+
+        target.set_name("test_target_name");
+        assert_eq!(source.name(), target.name());
+
+        let err = binding_group.bind("enabled", &target, "name").try_build();
+
+        assert!(err.is_err())
```
Comment (sdroege) [inline glib/src/gobject/binding_group.rs:289]: This fails because the binding group is **not** checking if the types are matching, and it also really can't because the transform functions might make it work for this set of types.

File: glib/src/gobject/binding_group.rs [hunk 4]
```diff
@@ -120,7 +120,10 @@ impl<'a> BindingGroupBuilder<'a> {
                 None => false,
                 Some(res) => {
                     assert!(
-                        res.type_().is_a(transform_data.3.value_type()),
+                        crate::Value::type_transformable(
+                            res.type_(),
+                            transform_data.3.value_type()
+                        ),
```
Comment (sdroege) [inline glib/src/gobject/binding_group.rs:126]: Why this change?

File: glib/src/gobject/binding_group.rs [hunk 5]
```diff
@@ -111,7 +111,7 @@ impl<'a> BindingGroupBuilder<'a> {
             user_data: ffi::gpointer,
         ) -> ffi::gboolean {
             let transform_data =
-                &*(user_data as *const (TransformFn, TransformFn, ParamSpec, ParamSpec));
```
Comment (sdroege) [inline glib/src/gobject/binding_group.rs:114]: It would be good to have a test with the transform functions to make sure this actually works properly.

File: glib/src/gobject/binding_group.rs [hunk 6]
```diff
@@ -140,19 +143,26 @@ impl<'a> BindingGroupBuilder<'a> {
...
 
             match (transform_data.1.as_ref().unwrap())(
-                &from_glib_borrow(binding),
+                &binding,
                 &*(from_value as *const Value),
             ) {
                 None => false,
                 Some(res) => {
+                    let pspec_name = transform_data.2.clone();
+                    let source = binding.source().unwrap();
```
Comment (sdroege) [inline glib/src/gobject/binding_group.rs:156]: This is not thread-safe I think. The source might be unset while the closure is running or not? CC @jf2048

Comment (jf2048) [reply glib/src/gobject/binding_group.rs:156]: Yes, it seems the real source is kept on the stack: https://gitlab.gnome.org/GNOME/glib/-/blob/91407cd7efec344f6bb5fd88fd7f771afb4fed25/gobject/gbinding.c

File: glib/src/gobject/binding_group.rs (hunk 4 above)
Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs:126]: I think the check is more correct, as it should accept something that is transformable to the actual type and not exactly the same one (it was what it it was suggested for the other closure too), but it needs a test

File: glib/src/gobject/binding_group.rs (hunk 6 above)
Comment (jf2048) [reply glib/src/gobject/binding_group.rs:156]: This should be thread safe actually. `GBinding:source` is construct only, and the notify handler will always hold a reference while the transform function is running

Comment (sdroege) [reply glib/src/gobject/binding_group.rs:156]: How is this construct-only if there's `binding_group.set_source(Some(&source))` (in the test)? :)

Comment (jf2048) [reply glib/src/gobject/binding_group.rs:156]: This is in the `GBinding`, not the `GBindingGroup`. The binding group creates a new `GBinding` for each property, each `GBinding` holds a construct only weak ref to the source.

Comment (sdroege) [reply glib/src/gobject/binding_group.rs:156]: Uh... right. Seems indeed fine then :)

File: glib/src/gobject/binding_group.rs (hunk 4 above)
Comment (sdroege) [reply glib/src/gobject/binding_group.rs:126]: Yes please. Is something actually going to do that transform at some point ;)

Also this probably means that `GBinding` bindings need to be updated the same way?

Comment (sdroege) [reply glib/src/gobject/binding_group.rs:126]: Also unclear if we want to allow that. This means that you can pass a `0u32` and it would become a `false`, for example. For properties/signals we disallow automatic conversions as that seems out of place in Rust.

Comment (jf2048) [reply glib/src/gobject/binding_group.rs:126]: Let's leave it as `type_().is_a`, that should disallow numeric conversions but still allow for object/interface types to be passed without casting them

**Comment by A6GibKm** [comment]
> @A6GibKm The problem was in your code: you were not NUL-terminating the string if there was no source object.
> 
> Now the test is still failing but that's something plainly on the Rust side of things :)

Is this Ok now? what does NULL-terminating a string even means in rust.

File: glib/src/gobject/binding_group.rs (hunk 4 above)
Comment (sdroege) [reply glib/src/gobject/binding_group.rs:126]: @A6GibKm This is still pending

File: glib/src/gobject/binding_group.rs (hunk 5 above)
Comment (sdroege) [reply glib/src/gobject/binding_group.rs:114]: @A6GibKm This is still pending

File: glib/src/gobject/binding_group.rs (hunk 3 above)
Comment (sdroege) [reply glib/src/gobject/binding_group.rs:289]: @A6GibKm This is still pending

**Comment by sdroege** [comment]
> Is this Ok now?

See comments I added to the still pending discussions above.

> what does NULL-terminating a string even means in rust.

These are strings that are provided to C functions. They need to have a 0-byte at the end.

File: glib/src/gobject/binding_group.rs (hunk 3 above)
Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs]: Removed the assert.

File: glib/src/gobject/binding_group.rs (hunk 4 above)
Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs]: Ok leaved as it was.

File: glib/src/gobject/binding_group.rs (hunk 5 above)
Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs:114]: Added tests but they fail here locally with 
```
Caused by:
  process didn't exit successfully: `/var/home/user/Projects/gtk-rs-core/target/debug/deps/glib-73fbdc0a479b0288` (signal: 11, SIGSEGV: invalid memory reference)
  ```

Comment (sdroege) [reply glib/src/gobject/binding_group.rs:114]: Good that we added a test then :)

Comment (jf2048) [reply glib/src/gobject/binding_group.rs:114]: I think the crash is because this should be `String, ParamSpec`?

Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs:114]: It is like that now

File: glib/src/gobject/binding_group.rs [hunk 7]
```diff
@@ -166,20 +173,30 @@ impl<'a> BindingGroupBuilder<'a> {
             let _ = Box::from_raw(data as *mut (TransformFn, TransformFn, ParamSpec, ParamSpec));
```
Comment (jf2048) [inline glib/src/gobject/binding_group.rs:173]:
```suggestion
            let _ = Box::from_raw(data as *mut (TransformFn, TransformFn, String, ParamSpec));
```

**Review by jf2048 — CHANGES_REQUESTED** [review]

File: glib/src/gobject/binding_group.rs (hunk 5 above)
Comment (jf2048) [reply glib/src/gobject/binding_group.rs:114]: Sorry not this one, it's in another spot. See my other review

File: glib/src/gobject/binding_group.rs (hunk 7 above)
Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs]: I think it does, but i have crashes locally on smething else

Comment (sdroege) [reply glib/src/gobject/binding_group.rs]: The problem there was that you pass a `*const u8` and not a `String`. And that `*const u8` would even be freed already in various cases.

File: glib/src/gobject/binding_group.rs [hunk 8]
```diff
@@ -140,19 +140,26 @@ impl<'a> BindingGroupBuilder<'a> {
...
                 Some(res) => {
+                    let pspec_name = transform_data.2.clone();
+                    let source = binding.source().unwrap();
+                    let pspec = source.find_property(&pspec_name);
+                    assert!(pspec.is_some(), "Source object does not have a property {}", pspec_name);
+                    let pspec = pspec.unwrap();
+
                     assert!(
-                        res.type_().is_a(transform_data.2.value_type()),
+                        crate::Value::type_transformable(res.type_(), pspec.value_type()),
```
Comment (sdroege) [inline glib/src/gobject/binding_group.rs:159]: This should stay as before. We don't want to loosen the type checks, as discussed.

File: glib/src/gobject/binding_group.rs [hunk 9]
```diff
@@ -233,8 +249,197 @@ impl<'a> BindingGroupBuilder<'a> {
...
+                Some(format!("{} World", value).to_value())
+            })
+            .build();
+
+        source.set_name("Hello");
+        assert_eq!(target.name(), "Hello World");
+    }
+
+    #[test]
+    fn binding_from_transform() {
```
Comment (sdroege) [inline glib/src/gobject/binding_group.rs:322]: Would be good to have transform tests where the source/target property have different types.

**Review by sdroege — COMMENTED** [review]
Apart from those two comments this should be OK now.

File: glib/src/gobject/binding_group.rs (hunk 9 above)
Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs:322]: give me an example of different types and ill code it, I tried using the current 2 properties (bool and string) and of course it was not possible due to the checks we do, even if we can do the conversion in the GtkExpression.

File: glib/src/gobject/binding_group.rs (hunk 8 above)
Comment (A6GibKm) [reply glib/src/gobject/binding_group.rs]: Done.

**Review by sdroege — APPROVED** [review]

File: glib/src/gobject/binding_group.rs (hunk 9 above)
Comment (sdroege) [reply glib/src/gobject/binding_group.rs:322]: See https://github.com/gtk-rs/gtk-rs-core/issues/700
